### PR TITLE
K8SPXC-1262: improve PITR

### PR DESCRIPTION
**Pull request by pooknull**
[K8SPXC-1262](https://jira.percona.com/browse/K8SPXC-1262) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

https://jira.percona.com/browse/K8SPXC-1262

**DESCRIPTION**
---
**Problem:**
`pitr` test fails because gap is detected in binary logs after restore, and the next restore is unable to start because it doesn't have `percona.com/unsafe-pitr: "true"` annotation. Gap shouldn't be detected in this test.

**Cause:**

1. The main problem is that the pitr collector can start while restore is running. During the restore it can't find the required binlog and throws a gap error.
2. Also, there are problems with finding last uploaded binlog. Currently, we just compare all binlog GTID sets with last uploaded binlog GTID set as two strings: https://github.com/percona/percona-xtradb-cluster-operator/blob/9072b5c77fcf695ac1439a14b9427bf38709f188/cmd/pitr/collector/collector.go#L263-L268
This solution has two problems:
   - GTID sets can have multiple GTIDs separated by commas, for example `2174B383-5441-11E8-B90A-C80AA9429562:1-3, 24DA167-0C0C-11E8-8442-00059A3C7B00:1-19`. We can have a problem when the present binlog can have only one GTID, while the last uploaded binlog GTID set can have multiple.
   - Last uploaded binlog GTIDs can have a transaction ID ranges, while present binlog can have only one transaction ID (`2174B383-5441-11E8-B90A-C80AA9429562:1-3` and `2174B383-5441-11E8-B90A-C80AA9429562:3` for example).
3.  We store only first source ID of GTID set in the storage for last uploaded binlog: https://github.com/percona/percona-xtradb-cluster-operator/blob/9072b5c77fcf695ac1439a14b9427bf38709f188/cmd/pitr/collector/collector.go#L390

During testing, `pitr` test got false positives. We are checking for gap errors in the `Reconcile` of `PerconaXtraDBCluster`, which is called almost every 5 seconds. If pitr collector detected a gap error, and restore is created before this `Reconcile` function is called, then restore is starting despite the gap error. During the restore, `PerconaXtraDBClusterBackup` recieves a `Binlog with GTID set XXX not found` condition

**Solution:**

1. Check if restore is running before creating pitr collector deployment
2. For comparing GTID sets, they should be divided into GTID arrays and compared separately. Also, we should use [GTID_SUBSET](https://dev.mysql.com/doc/refman/8.0/en/gtid-functions.html#function_gtid-subset) function for comparing GTIDs to resolve problem with different transaction ID ranges.
3. Files with `last-binlog-name-` prefix should be stored per GTID in the set. These files should contain the same content but have different GTID source IDs in the name.
4. We should check for pitr gap errors just before the restore

**CHECKLIST**
---
**Jira**
- [x] Is the Jira ticket created and referenced properly?
- [x] Does the Jira ticket have the proper statuses for documentation (`Needs Doc`) and QA (`Needs QA`)?
- [x] Does the Jira ticket link to the proper milestone (Fix Version field)?

**Tests**
- [ ] Is an E2E test/test case added for the new feature/change?
- [ ] Are unit tests added where appropriate?
- [ ] Are OpenShift compare files changed for E2E tests (`compare/*-oc.yml`)?

**Config/Logging/Testability**
- [ ] Are all needed new/changed options added to default YAML files?
- [ ] Are the manifests (crd/bundle) regenerated if needed?
- [ ] Did we add proper logging messages for operator actions?
- [ ] Did we ensure compatibility with the previous version or cluster upgrade process?
- [ ] Does the change support oldest and newest supported PXC version?
- [ ] Does the change support oldest and newest supported Kubernetes version?

### CODE REVIEW

File: cmd/pitr/collector/collector.go [hunk 1]
```diff
@@ -242,35 +242,53 @@ func (c *Collector) CollectBinLogs(ctx context.Context) error {
...
+		for i := len(list) - 1; i >= 0 && lastUploadedBinlogName == ""; i-- {
+			for _, gtidSet := range list[i].GTIDSet.List() {
+				for _, lastUploaded := range c.lastUploadedSet.List() {
+					isSubset, err := c.db.GTIDSubset(ctx, lastUploaded, gtidSet)
+					if err != nil {
+						return errors.Wrap(err, "check if gtid set is subset")
+					}
+					if isSubset {
+						lastUploadedBinlogName = list[i].Name
+						break
```
Comment (egegunes) [inline cmd/pitr/collector/collector.go:282]: this only break innermost for loop, do we want to continue if we found the binlog?

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| affinity-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/affinity-8-0.log) |
| auto-tuning-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/auto-tuning-8-0.log) |
| cross-site-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/cross-site-8-0.log) |
| demand-backup-cloud-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/demand-backup-cloud-8-0.log) |
| demand-backup-encrypted-with-tls-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/demand-backup-encrypted-with-tls-8-0.log) |
| demand-backup-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/demand-backup-8-0.log) |
| haproxy-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/haproxy-5-7.log) |
| haproxy-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/haproxy-8-0.log) |
| init-deploy-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/init-deploy-5-7.log) |
| init-deploy-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/init-deploy-8-0.log) |
| limits-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/limits-8-0.log) |
| monitoring-2-0-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/monitoring-2-0-8-0.log) |
| one-pod-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/one-pod-5-7.log) |
| one-pod-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/one-pod-8-0.log) |
| pitr-8-0 | [failure](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/pitr-8-0.log) |
| pitr-gap-errors-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/pitr-gap-errors-8-0.log) |
| proxy-protocol-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/proxy-protocol-8-0.log) |
| proxysql-sidecar-res-limits-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/proxysql-sidecar-res-limits-8-0.log) |
| recreate-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/recreate-8-0.log) |
| restore-to-encrypted-cluster-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/restore-to-encrypted-cluster-8-0.log) |
| scaling-proxysql-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/scaling-proxysql-8-0.log) |
| scaling-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/scaling-8-0.log) |
| scheduled-backup-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/scheduled-backup-5-7.log) |
| scheduled-backup-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/scheduled-backup-8-0.log) |
| security-context-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/security-context-8-0.log) |
| smart-update1-8-0 | [failure](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/smart-update1-8-0.log) |
| smart-update2-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/smart-update2-8-0.log) |
| storage-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/storage-8-0.log) |
| tls-issue-cert-manager-ref-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/tls-issue-cert-manager-ref-8-0.log) |
| tls-issue-cert-manager-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/tls-issue-cert-manager-8-0.log) |
| tls-issue-self-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/tls-issue-self-8-0.log) |
| upgrade-consistency-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/upgrade-consistency-8-0.log) |
| upgrade-haproxy-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/upgrade-haproxy-8-0.log) |
| upgrade-proxysql-8-0 | [failure](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/upgrade-proxysql-8-0.log) |
| users-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/users-5-7.log) |
| users-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/users-8-0.log) |
| validation-hook-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/validation-hook-8-0.log) |
| We run 37 out of 37|

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/1441/commits/caf50c7b2d131ff17c8281556ba8b308d0393763
image: `perconalab/percona-xtradb-cluster-operator:PR-1441-caf50c7b`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| affinity-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/affinity-8-0.log) |
| auto-tuning-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/auto-tuning-8-0.log) |
| cross-site-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/cross-site-8-0.log) |
| demand-backup-cloud-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/demand-backup-cloud-8-0.log) |
| demand-backup-encrypted-with-tls-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/demand-backup-encrypted-with-tls-8-0.log) |
| demand-backup-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/demand-backup-8-0.log) |
| haproxy-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/haproxy-5-7.log) |
| haproxy-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/haproxy-8-0.log) |
| init-deploy-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/init-deploy-5-7.log) |
| init-deploy-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/init-deploy-8-0.log) |
| limits-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/limits-8-0.log) |
| monitoring-2-0-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/monitoring-2-0-8-0.log) |
| one-pod-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/one-pod-5-7.log) |
| one-pod-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/one-pod-8-0.log) |
| pitr-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/pitr-8-0.log) |
| pitr-gap-errors-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/pitr-gap-errors-8-0.log) |
| proxy-protocol-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/proxy-protocol-8-0.log) |
| proxysql-sidecar-res-limits-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/proxysql-sidecar-res-limits-8-0.log) |
| recreate-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/recreate-8-0.log) |
| restore-to-encrypted-cluster-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/restore-to-encrypted-cluster-8-0.log) |
| scaling-proxysql-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/scaling-proxysql-8-0.log) |
| scaling-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/scaling-8-0.log) |
| scheduled-backup-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/scheduled-backup-5-7.log) |
| scheduled-backup-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/scheduled-backup-8-0.log) |
| security-context-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/security-context-8-0.log) |
| smart-update1-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/smart-update1-8-0.log) |
| smart-update2-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/smart-update2-8-0.log) |
| storage-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/storage-8-0.log) |
| tls-issue-cert-manager-ref-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/tls-issue-cert-manager-ref-8-0.log) |
| tls-issue-cert-manager-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/tls-issue-cert-manager-8-0.log) |
| tls-issue-self-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/tls-issue-self-8-0.log) |
| upgrade-consistency-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/upgrade-consistency-8-0.log) |
| upgrade-haproxy-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/upgrade-haproxy-8-0.log) |
| upgrade-proxysql-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/upgrade-proxysql-8-0.log) |
| users-5-7 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/users-5-7.log) |
| users-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/users-8-0.log) |
| validation-hook-8-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pxc-operator/PR-1441/caf50c7b/validation-hook-8-0.log) |
| We run 37 out of 37|

commit: https://github.com/percona/percona-xtradb-cluster-operator/pull/1441/commits/caf50c7b2d131ff17c8281556ba8b308d0393763
image: `perconalab/percona-xtradb-cluster-operator:PR-1441-caf50c7b`

File: cmd/pitr/collector/collector.go [hunk 2]
```diff
@@ -242,35 +242,53 @@ func (c *Collector) CollectBinLogs(ctx context.Context) error {
...
-	if c.lastSet != "" {
-		for i := len(list) - 1; i >= 0; i-- {
-			if list[i].GTIDSet == c.lastSet {
-				lastUploadedBinlogName = list[i].Name
-				break
+	if !c.lastUploadedSet.IsEmpty() {
+		for i := len(list) - 1; i >= 0 && lastUploadedBinlogName == ""; i-- {
+			for _, gtidSet := range list[i].GTIDSet.List() {
+				for _, lastUploaded := range c.lastUploadedSet.List() {
+					isSubset, err := c.db.GTIDSubset(ctx, lastUploaded, gtidSet)
```
Comment (egegunes) [inline cmd/pitr/collector/collector.go:276]: last uploaded GTID set can be superset of the current GTID set too. 

for example, collector was collecting from pod-0. pod-0 has the following binlogs:
```
binlog.000011 fbb06ce1-1579-11ee-902a-774be528cc87:1-352
binlog.000012 fbb06ce1-1579-11ee-902a-774be528cc87:353-553
binlog.000013 fbb06ce1-1579-11ee-902a-774be528cc87:554-754
binlog.000014 fbb06ce1-1579-11ee-902a-774be528cc87:755-957
binlog.000015 fbb06ce1-1579-11ee-902a-774be528cc87:958-1087
binlog.000016 fbb06ce1-1579-11ee-902a-774be528cc87:1088-1116
```

then I killed pod-0 and collector connected to pod-1 and uploaded the following binlog:
```
binlog.000014 fbb06ce1-1579-11ee-902a-774be528cc87:1-1185
```

after pod-0 comes back and collector connects to it, it checks if `binlog.000014 fbb06ce1-1579-11ee-902a-774be528cc87:1-1185` (last uploaded) is a subset of GTIDs in pod-0 but it's not, it's a superset.

Comment (pooknull) [reply cmd/pitr/collector/collector.go:279]: https://github.com/percona/percona-xtradb-cluster-operator/pull/1441/commits/8ce508d0bc168996f55fd5939dadab49bdf1b2e0

File: cmd/pitr/collector/collector.go (hunk 1 above)
Comment (pooknull) [reply cmd/pitr/collector/collector.go:285]: https://github.com/percona/percona-xtradb-cluster-operator/pull/1441/commits/8ce508d0bc168996f55fd5939dadab49bdf1b2e0

**Review by egegunes — APPROVED** [review]

**Review by hors — APPROVED** [review]

**Pull request merged**